modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql-extensions/src/main/java/org/nuxeo/ecm/core/storage/sql/db/EmbeddedFunctions.java

```diff
@@ -132,82 +143,93 @@ public static boolean isAccessAllowed(Serializable id, Set<String> principals, S
+            RowInfo rowInfo = null; // info about the row for the current id
+            if (disableVersionACL) {
+                // if it's a version, ignore its ACL and find the live doc
+                rowInfo = getRowInfo(psHier, psVer, id);
+                if (rowInfo.isVersion) {
+                    id = rowInfo.versionableId;
+                    if (id == null) {
+                        return false;
                     }
+                    rowInfo = null;
                 }
-                /*
-                 * Nothing conclusive found, repeat on the parent.
-                 */
-                ps2.setObject(1, id);
-                Serializable newId;
-                try (ResultSet rs = ps2.executeQuery()) {
-                    if (rs.next()) {
-                        newId = (Serializable) rs.getObject(1);
-                        if (rs.wasNull()) {
-                            newId = null;
-                        }
-                    } else {
-                        // no such id
-                        newId = null;
-                    }
+            }
+            do {
+                // check permissions at this level
+                Boolean access = getAccess(psAcl, id, principals, permissions);
+                if (access != null) {
+                    return access;
+                }
+                // nothing conclusive found, repeat on the parent
+                if (rowInfo == null) {
+                    rowInfo = getRowInfo(psHier, psVer, id);
+                }
+                if (rowInfo.isVersion) {
+                    id = rowInfo.versionableId;
+                } else {
+                    id = rowInfo.parentId;
                 }
-                if (first && newId == null) {
-                    // there is no parent for the first level
-                    // we may have a version on our hands, find the live doc
-                    try (PreparedStatement ps3 = conn.prepareStatement(
-                            "SELECT VERSIONABLEID FROM VERSIONS WHERE ID = ?")) {
-                        ps3.setObject(1, id);
-                        try (ResultSet rs = ps3.executeQuery()) {
-                            if (rs.next()) {
-                                newId = (Serializable) rs.getObject(1);
-                                if (rs.wasNull()) {
-                                    newId = null;
-                                }
-                            } else {
-                                // no such id
-                                newId = null;
-                            }
+                rowInfo = null;
+            } while (id != null);
+            // we've reached the root or a placeless document, deny access
+            return false;
+        }
+    }
+
+    protected static class RowInfo {
+
+        public Serializable parentId;
+
+        public boolean isVersion;
+
+        public Serializable versionableId;
```

Comment:
maybe make the members final as they are meant to be set once per instantiation ?

Reply:
That would mean writing a constructor, and not being able to update the fields independently which is how this is used today. So no, that would be much more verbosity.